src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+                );
+
+        BigDecimal total = shoppingCart.getCartItems().stream()
+                .map(
+                        cartItem -> cartItem
+                                .getBook()
+                                .getPrice()
+                                .multiply(BigDecimal.valueOf(cartItem.getQuantity()))
+                )
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setTotal(total);
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(requestDto.getShippingAddress());
+        orderRepository.save(order);
+
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(orderItemMapper::toOrderItem)
+                .peek(orderItem -> orderItem.setOrder(order))
+                .collect(Collectors.toSet());
+
+        order.setOrderItems(orderItems);
+        orderItemRepository.saveAll(orderItems);
+        cartItemRepository.deleteAll(shoppingCart.getCartItems());
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public List<OrderDto> getOrderHistory(Long userId) {
+        List<Order> orders = orderRepository.findByUserId(userId);
+        return orders.stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public OrderDto updateStatus(UpdateOrderStatusRequestDto requestDto, Long id) {
+        Order order = orderRepository.findById(id)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find order by id: " + id)
+                );
+        order.setStatus(requestDto.getStatus());
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public List<OrderItemDto> getOrderItems(Long userId, Long orderId) {
+        Order order = orderRepository.findByUserIdAndId(userId, orderId)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find order by id: " + orderId)
+                );
+        return order.getOrderItems().stream()
+                .map(orderItemMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public OrderItemDto getSpecificOrderItem(Long userId, Long itemId, Long orderId) {
+        Order order = orderRepository.findByUserIdAndId(userId, orderId)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find order by id: " + orderId)
+                );
+        OrderItem item = order.getOrderItems().stream()
+                .filter(orderItem -> orderItem.getId().equals(itemId))
+                .findFirst()
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find order item by id: " + itemId)
+                );
```

Comment:
The following method can be used to split this code, eliminating the need to create a query in repo.
`OrderItem item = orderItemRepository.findByIdAndOrderId(itemId, orderId)`

Reply:
agree